mason bash completion script

ankingcodes opened the pull request:
fixes #16014 

This PR aims to introduce a bash completion script for mason, which when loaded to shell using `source` can provide command completion for mason commands.
Commands included: 
- [x] mason new
- [x] mason init
- [x] mason add
- [x] mason rm
- [x] mason run
- [x] mason build
- [x] mason update
- [x] mason search
- [x] mason env
- [x] mason clean
- [x] mason doc
- [x] mason test
- [x] mason publish
- [x] mason external 
- [x] mason system

ankingcodes commented:
Currently, the script provides completion for `mason` command and all the sub-commands. For further auto-completion, I have only implemented for `mason new`, which shows the options available to `mason new`(package name depends on user, and directory has not been created before using command).  

@ben-albrecht @ronawho - requesting for an early review since this is going to be a long script considering the number of commands, thus, avoiding a long review time and large amount of requested changes.

ben-albrecht commented:
> For further auto-completion, I have only implemented for mason new, which shows the options available to mason new(package name depends on user, and directory has not been created before using command).

Since we're only addressing part of #16014, can you adjust the PR message to say that? The key phrase "fixes #16014" will auto-close that issue upon merging.

ankingcodes commented:
@ben-albrecht I'm planning to cover all commands in future commits so when this PR gets merge I'd assume the corresponding issue to be closed.

ankingcodes commented:
@ben-albrecht I added completion for `mason init`. Also, I ran a shellcheck on the script and I fixed the errors. However, to check for posix-compliance, I changed the shebang to `#!/bin/sh` and did a shellcheck again, and I found a bunch of errors for things which are allowed in bash, without which its really difficult to make this work. The following is part of the long list of error outputs :
```
In mason-completion.bash line 4:
local optional
 ^-- SC2039: In POSIX sh, 'local' is undefined.

In mason-completion.bash line 20:
COMPREPLY=($(compgen -W "$optional $optional_abbr" -- "$1"))
          ^-- SC2039: In POSIX sh, arrays are 
               ^-- SC2039: In POSIX sh, 'compgen' is undefined.    
```

ben-albrecht commented:
@ankingcodes - thanks for checking that. Do our other bash completion scripts do this? Perhaps we don't care about POSIX-friendliness in bash completion...  It is _bash_ completion and not _dash_ completion after all.

@ronawho - can you confirm whether or not we currently care about POSIX-compatibility for bash completion scripts?

ankingcodes commented:
@ben-albrecht I don't think we care for POSIX compatibility as per work done in https://github.com/chapel-lang/chapel/blob/master/util/devel/chpl-bash-completion-template

ben-albrecht commented:
@ankingcodes - perhaps we should switch this PR to a draft PR until it's completed and ready for a full review? The initial implementation still look good to me.

ankingcodes commented:
@ben-albrecht We could do that ! 
As of now, we are using a set of commands to parse `--` options from a command's help message, to ease maintenance of the script. This approach however causes problems for some commands such as `mason build` where the help message contains examples and usage of `--option` in between description of the command.
I am thinking of getting off of our approach  and store the options by hard-coding in our script. Do you have other ideas ?

ben-albrecht commented:
> As of now, we are using a set of commands to parse -- options from a command's help message, to ease maintenance of the script. This approach however causes problems for some commands such as mason build where the help message contains examples and usage of --option in between description of the command.
I am thinking of getting off of our approach and store the options by hard-coding in our script. Do you have other ideas ?

Could we introduce a regex pattern that essentially looks for:

- 1+ characters of leading white space at the beginning of a newline
- followed by a `-` character
- followed by 1+ non-whitespace characters

and use that to get the the flags more reliably?


That should be able to handle the `mason build -h` output:

```
Compile a local package and all of its dependencies

Usage:
    mason build [options]

Options:
    -h, --help                   Display this message
        --show                   Increase verbosity
        --release                Compile to target/release with optimizations (--fast)
        --force                  Force Mason to build the project
        --example <example>      Build an example from the example/ directory
        --[no-]update            [Do not] update the mason registry before building

When --example is thrown without an example, all examples will be built
When no options are provided, the following will take place:
   - Build from mason project if Mason.lock present

Compilation flags and arguments can be included after mason arguments.
To ensure compilation flags and mason arguments to not conflict, separate them with a
single dash(`-`). For example
   e.g. mason build --force - --savec tmpdir
```

ankingcodes commented:
Sounds great ! What about ` --[no-]update ` ?

ben-albrecht commented:
> Also, the tab completion is a bit slow since we are parsing it from the help message.


Hmm, that's interesting. Maybe we could use these parsing scripts to generate a cache file that is used for the completion :), but let's focus on a basic implementation first.

> What about --[no-]update ?

Yeah, this one will be tricky. I think shorthand flags will be more challenging as well. You could try to make the regex pattern more sophisticated, or just use the regex pattern described above to identify the tokens containing commands and then do some additional pattern-matching on those tokens.

ankingcodes commented:
@ben-albrecht Done with completions of all commands. I developed a workaround for `--[no-]update` and the issue mentioned in above [comments](https://github.com/chapel-lang/chapel/pull/16035#issuecomment-661082265). 

This is ready for a full review. Also, how would this script be automatically sourced when an user install Chapel ?

ben-albrecht commented:
> This is ready for a full review. 

Great, thanks!

> Also, how would this script be automatically sourced when an user install Chapel ?

It would not - it's up to the user to source it themselves. We should mention it in the documentation in the same place the documentation mentions the `chpl` completion (not sure where that is off the top of my head).

ben-albrecht commented:
Brainstorming - I wonder if this script could be extended to generate HTML documentation like [this](https://doc.rust-lang.org/cargo/reference/manifest.html) down the road.

ankingcodes commented:
> I wonder if this script could be extended to generate HTML documentation 

I don't think we should use this script for that purpose but rather there are many javascript frameworks which makes it very easy to generate & maintain documentation. I can take a look on those and let you know. As we will gradually move Mason to a separate repository and have a separate website, I think this would be a good approach. What do you think ?

ankingcodes commented:
> It would not - it's up to the user to source it themselves.

Couldn't we include it in our Makefile (or some other way) ? AFAIK, most CLI tools source their bash completion scripts during installation, eg, git

ben-albrecht commented:
> I don't think we should use this script for that purpose but rather there are many javascript frameworks which makes it very easy to generate & maintain documentation

Yeah, I agree. I meant to generate something like restructuredText which could be used to generate HTML docs via Sphinx like we do the rest of our documentation.

> Couldn't we include it in our Makefile (or some other way) ? AFAIK, most CLI tools source their bash completion scripts during installation, eg, git

The problem is you have to have the user source the script, which can't be accomplished from a Makefile or running a script - because that is unable to access the parent environment. Other tools require users to add something to their `.bashrc`, which is what we'll have to do as well.

ben-albrecht commented:
I see one issue with `mason update` tab completion outside of a mason project:

```sh
> mason update <tab>
Mason could not find your configuration file (Mason.toml)
```

ben-albrecht commented:
@ankingcodes - do you know if it's possible to not tab complete any flags unless the user already starts with a `-`? 

Right now, we get suggested flags when tab completing a command, e.g.

```
mason env <tab>
mason env - <tab><tab>
--help  -h
```

I think I'd be OK to keep this as is, but if it's possible to make it more consistent with other tab completions, that'd be ideal.

ankingcodes commented:
@ben-albrecht - This occurs only with commands which doesn't have an extra argument after that command name. In, that case I believe it's okay to complete the `-`(I don't know as of now if that can be disabled) , but for other commands such as `mason init` , we are shown directories, or files as required.

ben-albrecht commented:
@ankingcodes - OK. I think this is good barring the `mason update` issue then.

ankingcodes commented:
>  good barring the mason update issue 

@ben-albrecht - Could you clarify the `mason update` issue ?

ben-albrecht commented:
@ankingcodes - I don't think tab completion for `mason update` should generate an error like above if using outside of a mason package: [see above](https://github.com/chapel-lang/chapel/pull/16035#issuecomment-668812601)

ankingcodes commented:
@ben-albrecht - Made the change at e1b1d66, I don't think you'll get error on `mason update` now.

ben-albrecht commented on util/devel/mason-completion.bash:
```diff
@@ -0,0 +1,275 @@
+      'doc')
+        _complete_mason_doc "$cur"
+        ;;
+      'test')
```
This is showing up as `est` for some reason:

```
> mason <tab-tab>
add       clean     env       external  new       rm        search    update
build     doc       est       init      publish   run       system
```

ankingcodes replied:
Could you share a screenshot ? Its showing correct output on my machine 
![adw](https://user-images.githubusercontent.com/40923324/89541319-d9eb1400-d81b-11ea-9b90-b197484e96fe.PNG)

ben-albrecht replied:
Maybe it's an MacOS thing... I'll try to debug a bit. Let me know if you have any ideas.

![Screen Shot 2020-08-06 at 10 20 32 AM](https://user-images.githubusercontent.com/3595449/89543080-7e466980-d7ce-11ea-9941-62a423ca32a5.png)

ankingcodes replied:
- Try typing `mason test <tab>` and check if the completion works correctly. 
- Type `mason -h` and check if `test` is displayed correctly there, since the completions are parsed from there.

ben-albrecht replied:
Thanks. I think I identified the cause and have a suggested work-around [here](https://github.com/chapel-lang/chapel/pull/16035#discussion_r466454960)

ben-albrecht commented on util/devel/mason-completion.bash:
```diff
@@ -0,0 +1,275 @@
+  cur=${COMP_WORDS[COMP_CWORD]}
+  prev="${COMP_WORDS[COMP_CWORD-1]}"
+  arg=$(mason --help | grep -v '^[A-Za-z]' | grep -v '^.*--'  | \
+     grep -v '<command>' | grep -v '\[options\]'  | sed 's/^[ \t]*//' | \
```
The `\t` in `sed 's/^[ \t]*//' ` is removing `t` character in the MacOS (BSD) version of sed.

ben-albrecht replied:
Based on [this SO post](https://stackoverflow.com/a/38023442/6771423) I believe we can fix this by using ANSI-C quote style:

```suggestion
     grep -v '<command>' | grep -v '\[options\]'  | sed $'s/^[ \t]*//' | \
```

If this works on your machine, do you want to switch all the sed commands to this mode?

ankingcodes commented:
@ben-albrecht - sorry for the delay, made the changes. Let me know if it works for you.

ben-albrecht commented:
Another odd one I found in some testing:

```
$ mason external install
compiler   find       info       install    search     uninstall
```

I would expect tab completion for the flags

ben-albrecht commented:
@ankingcodes - I am seeing a similar issue as with `mason external` for `mason system pc`:

```
$ mason system pc
pc      search
```

All other subcommands worked as expected for me.

ankingcodes commented:
@ben-albrecht - made the changes.
